### PR TITLE
Integrate native three-stage vision alignment

### PR DESCRIPTION
**Pull request by RustinS**
# Integrate native three-stage vision alignment

## Summary

Add a maintained vision-alignment workflow on top of the native OLMoDDP infrastructure PR. The training entry point uses the existing internal experiment CLI and dataclass configuration system.

- Run bridge, perception, and joint training sequentially in one two-node allocation, with a fresh distributed process and optimizer for each phase. Explicit single-phase launches remain available.
- Build the LM architecture, tokenizer, and joint text replay from the pretrained checkpoint. Initialize the vision stack from its configured pretrained model.
- Use prepared visual sources, calibrated source weights, isolated packed-example attention, and resumable grouped text/vision loading. Source adapters remain configurable through ordinary dataset configs.
- Preserve model-only phase handoffs and full-state active-phase resume. A phase completes only after successful training and final-step held-out validation.
- Provide maintained fast-text, decoded-vision, and academic checkpoint-evaluation entry points with independently reusable completed work.
- Document configuration, source calibration, checkpoint handoffs, and evaluation separately from historical experiments.

## Default workflow

| Phase | Steps | Microbatch sequences/GPU | Trainable components |
| --- | ---: | ---: | --- |
| Bridge | 500 | 4 | Connector and reserved image-token input rows |
| Perception | 2,000 | 4 | Bridge components and vision encoder |
| Joint | 1,500 | 2 | Perception components and LM body; output projection frozen |

All phases use 8K sequences and a 64-crop packed-sequence limit. The per-image crop limit is separate. Bridge retains its 250-step cosine schedule followed by the learning-rate floor; perception and joint schedules match their training horizons. Frozen-LM phases disable router load-balancing loss; joint restores the pretrained per-layer coefficients. No router-input RMS repair is applied.

Joint text replay preserves the checkpoint's tokenizer, file ordering, filtering, masks, and storage dtype. Its held-out replay windows are excluded from joint training, not from the original LM pretraining.

## Evaluation compatibility

Saved evaluation panels remain inputs, not regenerated datasets. Decoded-panel deserialization adapts the former panel-class path without importing historical launch/config modules or changing row selections. Academic evaluation preserves saved recipients, donors, prompts, crop limits, generation limits, scoring, and overlap annotations. Those annotations describe the manifest's original training inventory, not every checkpoint evaluated later.

New result formats separate benchmark definitions from execution identity and do not adopt historical receipts as current completion records. Frozen historical deployments remain available outside this PR until distributed parity passes. The academic panel is a fixed validation subset, not the complete official benchmark splits.

## Scope and prerequisites

Depends on [#868](https://github.com/allenai/OLMo-core/pull/868), currently at `759e49820`. The prepared diff contains only vision model/data/training integration, the alignment recipe/controller, maintained evaluation components, focused tests, and documentation: 100 files, +22,580/-761 lines. Code/docs/config are 54 files (+11,731/-723); tests are 46 files (+10,849/-38).

Excluded: mixed-midtraining recipes and ablations, historical promotion/source-replay frameworks, experiment launch catalogs, frozen runtime archives, and personal result artifacts. Existing checkpoints with experiment-specific callback/config classes still require their matching frozen deployment; this PR does not rewrite those checkpoints.

## Validation and remaining merge gates

The isolated candidate currently passes 97 focused evaluation CPU tests without skips,
including checkpoint metadata/loading contracts, bounded image pairing, decoded-panel
compatibility, full academic saved-result rescoring, and prompt-token checks. Its complete
Sphinx dummy build succeeds with zero documentation warnings (58 pages; intersphinx network
lookups disabled), and all evaluation modules/tests pass Ruff, Black, and isort checks.

- [x] Validate the isolated candidate against the infrastructure branch, including imports, wheel contents, and documentation links.
- [x] Run focused vision model, dataset, packing, loss, replay, callback, controller, and checkpoint tests in that candidate.
- [ ] Confirm fresh-process all-stage handoffs, phase completion, and full-state restart in a distributed canary using the final candidate source.
- [ ] Complete legacy/native EP8 decoded and academic parity for the historical perception and current joint checkpoints.
- [x] Run changed-file style checks and inspect the final diff for historical dependencies or experiment files.

After the infra compatibility correction, the cross-layer suite passed 454 tests with five GPU-only checks skipped. The data suite separately passed 441 tests, and NumPy/replay allocation checks passed 21. Removing six unused annotation-hash methods passed 157 focused tests (overlapping the data suite). Real fresh-process Gloo controller checks passed; the final GPU training/resume canary remains outstanding. The wheel matches all 307 Python sources, imports directly without installation, and contains no generated binaries or legacy/mixed recipe artifacts. The source-only patch applies cleanly to an independent checkout of the exact stack base. These invocations overlap and are not summed into a unique total.

Candidate academic tests rescore 3,072 examples / 9,216 image-control scores and check saved
prompt token counts and generated-token decoding. Exact decoded rescoring of 896 saved rows
was additionally checked in the development worktree. Distributed evaluation parity is still
pending; no quality or speed improvement is claimed by this refactor.

### CODE REVIEW

File: src/olmo_core/eval/vision_academic.py [hunk 1]
```diff
@@ -0,0 +1,1621 @@
...
+                raise ValueError("Every academic option letter must encode to one response token")
+            input_ids, token_type_ids, position_ids = self._inputs(
+                prompt_ids,
+                required=len(prompt_ids),
+            )
+            logits_position = torch.tensor(
+                [[len(prompt_ids) - 1]], dtype=torch.long, device=self.device
+            )
+            with torch.inference_mode():
+                logits = self.train_module.model_forward_no_pipeline(
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/eval/vision_academic.py]: Addressed in 092958611 (included in current head 90400c72d). Image encoding and both LM prediction paths now use the existing eager, grad-enabled evaluation context, with enclosing inference mode disabled. Projected image features and logits are detached promptly. Tests cover MC/free-answer paths under normal, no-grad, and inference-mode callers, graph release, feature reuse, and mode restoration. An execution.forward_context field uses existing receipt equality checks to reject pre-fix cached fragments without changing benchmark definitions. The integrated 95-test suite passes, including saved-reference rescoring and prompt-token checks. These checks do not establish GPU prediction parity; that remains a separate pre-merge validation.

File: src/olmo_core/eval/vision_academic.py [hunk 2]
```diff
@@ -0,0 +1,1621 @@
...
+                    input_ids,
+                    encoded_image_features=encoded_features,
+                    token_type_ids=token_type_ids,
+                    position_ids=position_ids,
+                    logits_to_keep=logits_position,
+                )
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/eval/vision_academic.py:961]: The missing mask is observable, but the stated capacity-drop mechanism does not apply to this evaluator's supported path: _evaluate explicitly selects sync_1d and eval mode; OLMoDDPBlock uses the synchronized dropless path, which dispatches all N * top_k routes using exact counts. That dispatch also does not consume router_token_mask, so adding the argument alone would not change routing or padding overhead. Routing is left unchanged here. Mask/capacity handling would need revisiting if a capacity-limited evaluation backend is introduced.

File: src/olmo_core/data/numpy_dataset.py [hunk 3]
```diff
@@ -2580,6 +2583,8 @@ def build(self) -> NumpyDatasetBase:
             mixture = self.source_mixture_config.build(
                 npdtype=self.get_dtype(), sequence_length=self.sequence_length
             )
+            # Keep path and allocation indices aligned after dropping empty sources.
+            mixture.filter_zero_token_paths = True
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/data/numpy_dataset.py]: Addressed in 092958611 (included in current head 90400c72d). NumpyFSLDatasetConfig retains zero-allocation occurrences and their per-path metadata. Replay accepts the exact full saved path list or its exact positive-allocation projection, preserving order and duplicate occurrences; changed positive order/count/content is still rejected. Tiny real-data regressions cover both checkpoint representations, masks, metadata, zero-only index skipping, and shared-index sizing. Replay and mixture tests pass. The current s002-step125500 checkpoint uses the unweighted mix branch and was not affected.

File: src/olmo_core/train/callbacks/restore_metrics.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    priority: ClassVar[int] = 4
+
+    def pre_train(self):
+        """Merge the existing same-step file into the metric saver on rank zero."""
+        if get_rank() != 0 or not self.trainer.checkpoint_loaded:
+            return
+        callback = self.trainer.callbacks[self.metrics_callback]
+        path = Path(self.trainer.save_folder) / callback.step_metrics_fname.format(step=self.step)
+        if path.is_file():
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/train/callbacks/restore_metrics.py]: Addressed in 092958611 (included in current head 90400c72d). Restoration now checks file_exists on the intact URL and reads through resource_path instead of converting object-store URLs to pathlib paths. Tests verify local/S3/GCS restoration preserves training metrics when same-step evaluation updates the snapshot, plus rank-zero/checkpoint-loaded/missing-file guards. The metric-saver suite passes all 17 tests.

File: src/olmo_core/data/multimodal/finevision.py [hunk 5]
```diff
@@ -197,32 +251,142 @@ def _build_index(self) -> Optional[np.ndarray]:
         )
         return index
 
+    def _build_content_fingerprint(self) -> str:
+        """Build a stable identity for source rows, filtering, and serialization policy."""
+        cfg = self.config
+        arrow_fingerprint = (
+            cfg.expected_materialized_fingerprint
+            if cfg.expected_materialized_fingerprint is not None
+            else getattr(self._data, "_fingerprint", None)
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/data/multimodal/finevision.py:360]: Addressed in abc3f4674affa502175996dd5ba537635e4e7ab0, integrated at c6c884edd274cf6505e022516796b3bb3f8841b8. The underlying trust gap was real, but the configured pins are materialization-content SHA-256 values, not HF dataset fingerprints; HF fingerprints also change on column projection. The loader now recomputes the custom pin from its materialization manifest, checks the exact ordered Arrow files actually loaded and row totals, and verifies Arrow/dataset-info bytes against the pinned digests. Successful byte checks reuse the existing filesystem-cache mechanism under a dedicated data-verification environment variable; keys include canonical path, device/inode, size, nanosecond mtime/ctime, and expected digest. Changed files are rehashed, failures are not cached, and pre-load/post-verification identities detect replacements while mmap loading or verifying. A supplied pin is checked even without strict_annotations. Valid old resume identities remain unchanged. Tests cover projection, incorrect pins/inventories, same-size edits with restored mtime, cache hits/failures, and mutation races. The real Geo170K materialization passed verification and a no-rehash repeat; no full corpus replay was performed. The integrated 386-test PR suite passes.

File: src/olmo_core/data/multimodal/sft_common.py [hunk 6]
```diff
@@ -160,34 +208,124 @@ def decode_pil_image(obj: Any):
...
-            "max_sequence_length too small: truncation would drop <im_patch> tokens "
-            "(the image block must fit entirely within the sequence)."
+            "max_sequence_length too small: truncation would drop <im_patch> or other "
+            "image-structural tokens (the complete image block must fit within the sequence)."
         )
     if not np.any(seq["loss_masks"][:max_len] > 0):
         raise ValueError(f"Truncation to {max_len} removed all loss tokens")
     n = len(seq["input_ids"])
-    return {k: (v[:max_len] if v.ndim == 1 and len(v) == n else v) for k, v in seq.items()}
+    out = {k: (v[:max_len] if v.ndim == 1 and len(v) == n else v) for k, v in seq.items()}
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/data/multimodal/sft_common.py]: Addressed in bb051c9f1, integrated at efb6086aa55d7a09b0d76eb50c4f2efe0d4a7e75. Truncation now inspects ndim and slices only NumPy arrays; metadata dictionaries and other non-token values are preserved. The helper and MaxSequenceLengthDataset return annotations now allow those fields. Four parameterized cases cover direct/wrapped calls and truncation/no-op paths, checking token slices, metadata/scalars/lists/images/pooled indices, and unchanged input examples. Existing image-structural-token, all-loss-token, and branch-weight safeguards remain covered. All 61 formatter/dataset tests and the integrated 319-test PR suite pass; the corresponding original vision-moe verification passes 361 tests.

File: src/olmo_core/eval/multimodal_panel.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            [row["panel_index"] for row in rows] != list(range(examples))
+            or any(not isinstance(index, int) or index < 0 for index in indices)
+            or len(set(indices)) != examples
+        ):
+            raise ValueError(f"Frozen panel {name} has invalid or duplicate rows")
+        if name in OCR_COMPONENTS and (
+            not isinstance(source, SavedPanelSourceConfig)
+            or source.row_indices != indices
+            or source.added_validation_indices
+            != [row["base_source_index"] for row in rows if row["added_to_saved_panel"]]
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/eval/multimodal_panel.py:112]: Fixed in acf2e8b52, integrated with the infra update at df4cbeeba. Lightweight ordered selection checks now cover every source, including selection_repeat mapping, before model construction or cached-result reuse. Built dataset mappings and cached rank/complete row identities are checked against the frozen panel too. Metadata-only dry runs, benchmark definitions, prompts, scoring, and receipt schema remain unchanged; no new hashes or dataset replay were added. The 82 focused tests cover stale/reordered non-OCR manifests, changed selection files, cached row corruption, OCR safeguards, repeat mapping, and valid cache reuse. The combined stacked-PR CPU suite passes 352 tests (7 skipped, 3 GPU tests deselected); formatting and lint pass. The new checks also accepted the real frozen selections and all 1,344 saved rows across proper1500, mixed1x500, and mixed16x500 without loading datasets or models.

File: src/olmo_core/nn/vision/molmo2_tokens.py [hunk 8]
```diff
@@ -35,16 +40,124 @@
...
+        token_ids.im_end_id,
+        token_ids.im_patch_id,
+        token_ids.im_col_id,
+        token_ids.low_res_im_start_id,
+        token_ids.image_placeholder_id,
+        token_ids.im_end_turn_id,
+    ]
+    if len(set(resolved_ids)) != len(resolved_ids):
+        raise ValueError(f"Molmo2 special tokens did not resolve to unique IDs: {resolved_ids}")
+    if model_vocab_size is not None and max(resolved_ids) >= model_vocab_size:
```
[the remark this answers is not in the document]
Comment (RustinS) [reply src/olmo_core/nn/vision/molmo2_tokens.py]: Fixed in d43bf7062, integrated with the infra configuration fixes at 0647b2749. The bound check now uses the maximum ID across the entire loaded vocabulary after image-token additions, preserving required-token and special-ID uniqueness checks. Tests cover pre-existing/new image specials, sparse unrelated vocabulary IDs, the last valid embedding row, out-of-range IDs, and special-ID collisions. All 10 focused tests pass; the combined stacked-PR CPU suite passes 496 tests (3 GPU tests deselected). The updated function also accepts our pinned Dolma2 tokenizer offline unchanged: maximum adapted ID 100283 within 100352 model rows. Formatting and lint pass.
